api_tests/lib/bitcoin.ts

```diff
@@ -0,0 +1,156 @@
+import {
+    Transaction,
+    ECPair,
+    Out,
+    payments,
+    networks,
+    TransactionBuilder,
+    address,
+} from "bitcoinjs-lib";
+const BitcoinRpcClient = require("bitcoin-core");
+const sb = require("satoshi-bitcoin");
+const util = require("./util.js");
+
+interface IBitcoinConfig {
+    // snake_case because it comes from TOML file
+    rpc_username: string;
+    rpc_password: string;
+    rpc_host: string;
+    rpc_port: number;
+}
+
+interface IBitcoinRpcClient {
+    //TODO: do not use 'any'
+    generate(num: number): Promise<any>;
+    getRawTransaction(
+        txId: string,
+        verbose?: boolean,
+        blockHash?: string
+    ): Promise<any>;
+    sendToAddress(address: string, amount: number | string): Promise<any>;
+    sendRawTransaction(hexString: string): Promise<any>;
+}
+
+interface IUtxo {
+    // TODO: declare transaction Id type
+    txId: string;
+    value: number;
+    vout: number;
+}
+
+let _bitcoinRpcClient: IBitcoinRpcClient;
+let _bitcoinConfig: IBitcoinConfig;
+
+function createBitcoinRpcClient(btcConfig: IBitcoinConfig) {
+    if (!btcConfig && !_bitcoinConfig) {
+        throw new Error("bitcoin configuration is needed");
+    }
+
+    if (!_bitcoinRpcClient || btcConfig !== _bitcoinConfig) {
+        _bitcoinRpcClient = new BitcoinRpcClient({
+            network: "regtest",
+            port: btcConfig.rpc_port,
+            host: btcConfig.rpc_host,
+            username: btcConfig.rpc_username,
+            password: btcConfig.rpc_password,
+        });
+        _bitcoinConfig = btcConfig;
+    }
+    return _bitcoinRpcClient;
+}
+
+module.exports.createClient = (btcConfig: IBitcoinConfig) => {
+    return createBitcoinRpcClient(btcConfig);
+};
+
+module.exports.generate = async function(num: number = 1) {
+    return createBitcoinRpcClient(_bitcoinConfig).generate(num);
+};
+
+module.exports.activateSegwit = async function() {
+    return createBitcoinRpcClient(_bitcoinConfig).generate(432);
+};
+
+async function getFirstUtxoValueTransferredTo(txId: string, address: string) {
+    let satoshi = 0;
+    let tx = await _bitcoinRpcClient.getRawTransaction(txId, true);
+    let vout = tx.vout[0];
+
+    if (
+        vout.scriptPubKey.addresses.length === 1 &&
+        vout.scriptPubKey.addresses[0] === address
+    ) {
+        satoshi = sb.toSatoshi(vout.value);
+    }
+
+    return satoshi;
+}
+
+module.exports.get_first_utxo_value_transferred_to = getFirstUtxoValueTransferredTo;
+
+class BitcoinWallet {
+    keypair: ECPair;
+    bitcoinUtxos: IUtxo[];
+    _identity: {
+        address: string;
+        hash: Buffer;
+        output: Buffer;
+        pubkey: Buffer;
+        signature: Buffer;
+        input: Buffer;
+        witness: Buffer[];
+    };
+
+    constructor() {
+        this.keypair = ECPair.makeRandom({ rng: util.test_rng });
+        // TODO: Use wallet instead of array to track Bitcoin UTXOs
+        this.bitcoinUtxos = [];
+        this._identity = payments.p2wpkh({
```

Comment:
:question: Is there no proper return type for `p2wpkh`?

Reply:
Nope.